crates/networking/rpc/engine/fork_choice.rs

```diff
@@ -65,94 +70,148 @@ impl RpcHandler for ForkChoiceUpdatedV3 {
             self.fork_choice_state.finalized_block_hash
         );
 
-        let head_block = match apply_fork_choice(
-            &context.storage,
-            self.fork_choice_state.head_block_hash,
-            self.fork_choice_state.safe_block_hash,
-            self.fork_choice_state.finalized_block_hash,
-        ) {
-            Ok(head) => head,
-            Err(error) => {
-                let fork_choice_response = match error {
-                    InvalidForkChoice::NewHeadAlreadyCanonical => {
-                        ForkChoiceResponse::from(PayloadStatus::valid_with_hash(
-                            latest_canonical_block_hash(&context.storage).unwrap(),
-                        ))
-                    }
-                    InvalidForkChoice::Syncing => {
-                        // Start sync
-                        let current_number = context.storage.get_latest_block_number()?.unwrap();
-                        let Some(current_head) =
-                            context.storage.get_canonical_block_hash(current_number)?
-                        else {
-                            return Err(RpcErr::Internal(
-                                "Missing latest canonical block".to_owned(),
-                            ));
-                        };
-                        let sync_head = self.fork_choice_state.head_block_hash;
-                        tokio::spawn(async move {
-                            // If we can't get hold of the syncer, then it means that there is an active sync in process
-                            if let Ok(mut syncer) = context.syncer.try_lock() {
-                                syncer
-                                    .start_sync(current_head, sync_head, context.storage.clone())
-                                    .await
-                            }
-                        });
-                        ForkChoiceResponse::from(PayloadStatus::syncing())
-                    }
-                    reason => {
-                        warn!("Invalid fork choice state. Reason: {:#?}", reason);
-                        return Err(RpcErr::InvalidForkChoiceState(reason.to_string()));
-                    }
-                };
-                return serde_json::to_value(fork_choice_response)
-                    .map_err(|error| RpcErr::Internal(error.to_string()));
-            }
-        };
+        let (head_block_opt, mut response) =
+            handle_forkchoice(&self.fork_choice_state, context.clone())?;
```

Comment:
Most of the code has been extracted to `handle_forkchoice`, which we can reuse across the different FCU versions